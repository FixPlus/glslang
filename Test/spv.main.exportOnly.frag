#version 460 core

float foo(float, float);

void bar(inout vec3 v, int n)
{
    v.x += n;
}

int bar(vec3 v)
{
    return int(v.x + v.y + v.z);
}

float foo(int n, float s)
{
    return s * float(n);
}

out float cout;

void main()
{
    vec3 v = vec3(0, 1, 2);
    bar(v, -1);
    int n = bar(v);
    float x = foo(1.f, 2.f);
    float y = foo(n, 2.f);
    cout = x + y;
}

float foo(float a, float b)
{
    return a + b;
}
