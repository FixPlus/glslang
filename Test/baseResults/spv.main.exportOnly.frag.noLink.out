spv.main.exportOnly.frag
Shader version: 460
gl_FragCoord origin is upper left
0:? Sequence
0:5  Function Definition: bar(vf3;i1; ( global void)
0:5    Function Parameters: 
0:5      'v' ( inout highp 3-component vector of float)
0:5      'n' ( in highp int)
0:7    Sequence
0:7      add second child into first child ( temp highp float)
0:7        direct index ( temp highp float)
0:7          'v' ( inout highp 3-component vector of float)
0:7          Constant:
0:7            0 (const int)
0:7        Convert int to float ( temp highp float)
0:7          'n' ( in highp int)
0:10  Function Definition: bar(vf3; ( global highp int)
0:10    Function Parameters: 
0:10      'v' ( in highp 3-component vector of float)
0:12    Sequence
0:12      Branch: Return with expression
0:12        Convert float to int ( temp int)
0:12          add ( temp highp float)
0:12            add ( temp highp float)
0:12              direct index ( temp highp float)
0:12                'v' ( in highp 3-component vector of float)
0:12                Constant:
0:12                  0 (const int)
0:12              direct index ( temp highp float)
0:12                'v' ( in highp 3-component vector of float)
0:12                Constant:
0:12                  1 (const int)
0:12            direct index ( temp highp float)
0:12              'v' ( in highp 3-component vector of float)
0:12              Constant:
0:12                2 (const int)
0:15  Function Definition: foo(i1;f1; ( global highp float)
0:15    Function Parameters: 
0:15      'n' ( in highp int)
0:15      's' ( in highp float)
0:17    Sequence
0:17      Branch: Return with expression
0:17        component-wise multiply ( temp highp float)
0:17          's' ( in highp float)
0:17          Convert int to float ( temp highp float)
0:17            'n' ( in highp int)
0:22  Function Definition: main( ( global void)
0:22    Function Parameters: 
0:24    Sequence
0:24      Sequence
0:24        move second child to first child ( temp highp 3-component vector of float)
0:24          'v' ( temp highp 3-component vector of float)
0:24          Constant:
0:24            0.000000
0:24            1.000000
0:24            2.000000
0:25      Function Call: bar(vf3;i1; ( global void)
0:25        'v' ( temp highp 3-component vector of float)
0:25        Constant:
0:25          -1 (const int)
0:26      Sequence
0:26        move second child to first child ( temp highp int)
0:26          'n' ( temp highp int)
0:26          Function Call: bar(vf3; ( global highp int)
0:26            'v' ( temp highp 3-component vector of float)
0:27      Sequence
0:27        move second child to first child ( temp highp float)
0:27          'x' ( temp highp float)
0:27          Function Call: foo(f1;f1; ( global highp float)
0:27            Constant:
0:27              1.000000
0:27            Constant:
0:27              2.000000
0:28      Sequence
0:28        move second child to first child ( temp highp float)
0:28          'y' ( temp highp float)
0:28          Function Call: foo(i1;f1; ( global highp float)
0:28            'n' ( temp highp int)
0:28            Constant:
0:28              2.000000
0:29      move second child to first child ( temp highp float)
0:29        'cout' ( out highp float)
0:29        add ( temp highp float)
0:29          'x' ( temp highp float)
0:29          'y' ( temp highp float)
0:32  Function Definition: foo(f1;f1; ( global highp float)
0:32    Function Parameters: 
0:32      'a' ( in highp float)
0:32      'b' ( in highp float)
0:34    Sequence
0:34      Branch: Return with expression
0:34        add ( temp highp float)
0:34          'a' ( in highp float)
0:34          'b' ( in highp float)
0:?   Linker Objects
0:?     Linker Function: foo(f1;f1; ( global highp float)
0:?       Function Parameters: 
0:?         '' ( in highp float)
0:?         '' ( in highp float)
0:?     Linker Function: bar(vf3;i1; ( global void)
0:?       Function Parameters: 
0:?         '' ( inout highp 3-component vector of float)
0:?         '' ( in highp int)
0:?     Linker Function: bar(vf3; ( global highp int)
0:?       Function Parameters: 
0:?         '' ( in highp 3-component vector of float)
0:?     Linker Function: foo(i1;f1; ( global highp float)
0:?       Function Parameters: 
0:?         '' ( in highp int)
0:?         '' ( in highp float)
0:?     'cout' ( out highp float)

Validation failed
// Module Version 10000
// Generated by (magic number): 80008
// Id's are bound by 92

                              Capability Shader
                              Capability Linkage
               1:             ExtInstImport  "GLSL.std.450"
                              MemoryModel Logical GLSL450
                              EntryPoint Fragment 4  "main" 83
                              ExecutionMode 4 OriginUpperLeft
                              Source GLSL 460
                              Name 4  "main"
                              Name 14  "bar(vf3;i1;"
                              Name 12  "v"
                              Name 13  "n"
                              Name 18  "bar(vf3;"
                              Name 17  "v"
                              Name 24  "foo(i1;f1;"
                              Name 22  "n"
                              Name 23  "s"
                              Name 29  "foo(f1;f1;"
                              Name 27  "a"
                              Name 28  "b"
                              Name 58  "v"
                              Name 64  "param"
                              Name 66  "param"
                              Name 69  "n"
                              Name 70  "param"
                              Name 73  "x"
                              Name 74  "param"
                              Name 75  "param"
                              Name 77  "y"
                              Name 78  "param"
                              Name 80  "param"
                              Name 83  "cout"
                              Decorate 29(foo(f1;f1;) Linkage Attributes 678391654 1715155302 15153 0
                              Decorate 14(bar(vf3;i1;) Linkage Attributes 678584674 993224310 3879273 0
                              Decorate 18(bar(vf3;) Linkage Attributes 678584674 993224310 0 0
                              Decorate 24(foo(i1;f1;) Linkage Attributes 678391654 1715155305 15153 0
               2:             TypeVoid
               3:             TypeFunction 2
               6:             TypeFloat 32
               7:             TypeVector 6(float) 3
               8:             TypePointer Function 7(fvec3)
               9:             TypeInt 32 1
              10:             TypePointer Function 9(int)
              11:             TypeFunction 2 8(ptr) 10(ptr)
              16:             TypeFunction 9(int) 8(ptr)
              20:             TypePointer Function 6(float)
              21:             TypeFunction 6(float) 10(ptr) 20(ptr)
              26:             TypeFunction 6(float) 20(ptr) 20(ptr)
              33:             TypeInt 32 0
              34:     33(int) Constant 0
              41:     33(int) Constant 1
              45:     33(int) Constant 2
              59:    6(float) Constant 0
              60:    6(float) Constant 1065353216
              61:    6(float) Constant 1073741824
              62:    7(fvec3) ConstantComposite 59 60 61
              63:      9(int) Constant 4294967295
              82:             TypePointer Output 6(float)
        83(cout):     82(ptr) Variable Output
         4(main):           2 Function None 3
               5:             Label
           58(v):      8(ptr) Variable Function
       64(param):      8(ptr) Variable Function
       66(param):     10(ptr) Variable Function
           69(n):     10(ptr) Variable Function
       70(param):      8(ptr) Variable Function
           73(x):     20(ptr) Variable Function
       74(param):     20(ptr) Variable Function
       75(param):     20(ptr) Variable Function
           77(y):     20(ptr) Variable Function
       78(param):     10(ptr) Variable Function
       80(param):     20(ptr) Variable Function
                              Store 58(v) 62
              65:    7(fvec3) Load 58(v)
                              Store 64(param) 65
                              Store 66(param) 63
              67:           2 FunctionCall 14(bar(vf3;i1;) 64(param) 66(param)
              68:    7(fvec3) Load 64(param)
                              Store 58(v) 68
              71:    7(fvec3) Load 58(v)
                              Store 70(param) 71
              72:      9(int) FunctionCall 18(bar(vf3;) 70(param)
                              Store 69(n) 72
                              Store 74(param) 60
                              Store 75(param) 61
              76:    6(float) FunctionCall 29(foo(f1;f1;) 74(param) 75(param)
                              Store 73(x) 76
              79:      9(int) Load 69(n)
                              Store 78(param) 79
                              Store 80(param) 61
              81:    6(float) FunctionCall 24(foo(i1;f1;) 78(param) 80(param)
                              Store 77(y) 81
              84:    6(float) Load 73(x)
              85:    6(float) Load 77(y)
              86:    6(float) FAdd 84 85
                              Store 83(cout) 86
                              Return
                              FunctionEnd
 14(bar(vf3;i1;):           2 Function None 11
           12(v):      8(ptr) FunctionParameter
           13(n):     10(ptr) FunctionParameter
              15:             Label
              31:      9(int) Load 13(n)
              32:    6(float) ConvertSToF 31
              35:     20(ptr) AccessChain 12(v) 34
              36:    6(float) Load 35
              37:    6(float) FAdd 36 32
              38:     20(ptr) AccessChain 12(v) 34
                              Store 38 37
                              Return
                              FunctionEnd
    18(bar(vf3;):      9(int) Function None 16
           17(v):      8(ptr) FunctionParameter
              19:             Label
              39:     20(ptr) AccessChain 17(v) 34
              40:    6(float) Load 39
              42:     20(ptr) AccessChain 17(v) 41
              43:    6(float) Load 42
              44:    6(float) FAdd 40 43
              46:     20(ptr) AccessChain 17(v) 45
              47:    6(float) Load 46
              48:    6(float) FAdd 44 47
              49:      9(int) ConvertFToS 48
                              ReturnValue 49
                              FunctionEnd
  24(foo(i1;f1;):    6(float) Function None 21
           22(n):     10(ptr) FunctionParameter
           23(s):     20(ptr) FunctionParameter
              25:             Label
              52:    6(float) Load 23(s)
              53:      9(int) Load 22(n)
              54:    6(float) ConvertSToF 53
              55:    6(float) FMul 52 54
                              ReturnValue 55
                              FunctionEnd
  29(foo(f1;f1;):    6(float) Function None 26
           27(a):     20(ptr) FunctionParameter
           28(b):     20(ptr) FunctionParameter
              30:             Label
              87:    6(float) Load 27(a)
              88:    6(float) Load 28(b)
              89:    6(float) FAdd 87 88
                              ReturnValue 89
                              FunctionEnd
